### PR TITLE
add rotate by date

### PR DESCRIPTION
**Pull request by jb-alvarado**
Hi,
this is an idea of how to rotate by date (#12). How do you think about?

Line 388 needs still a correct assert.

### CODE REVIEW

**Comment by Ploppz** [comment]
Thanks for this PR.
The implementation looks mostly fine, I'll leave some comments on the code.

One thing I want to consider: Can we generalize it? Like `ContentLimit::Date(Duration)`. But maybe that's for the future.

File: src/lib.rs [hunk 1]
```diff
@@ -522,6 +540,15 @@ impl<S: SuffixScheme> FileRotate<S> {
 
         self.file = Some(File::create(&self.basepath)?);
 
+        if let Some(_) = &self.file {
```
Comment (Ploppz) [inline src/lib.rs:543]: the previous line says that `self.file = Some(_)`, so this check is superfluous

File: src/lib.rs [hunk 2]
```diff
@@ -522,6 +540,15 @@ impl<S: SuffixScheme> FileRotate<S> {
...
+            let metadata = metadata(&self.basepath).unwrap();
```
Comment (Ploppz) [inline src/lib.rs:544]: try to avoid unwrap. You can remove this line and change the next line to something like:
```
if let Ok(time) = metadata(&self.basepath).and_then(|metadta| metadata.modified()) {
```

File: src/lib.rs [hunk 3]
```diff
@@ -593,6 +620,18 @@ impl<S: SuffixScheme> Write for FileRotate<S> {
                     file.write_all(buf)?;
                 }
             }
+            ContentLimit::Date => {
+                let local: DateTime<Local> = Local::now();
+                if let Some(modified) = self.modified {
```
Comment (Ploppz) [inline src/lib.rs:625]: Is it generally the assumption that if we have `ContentLimit::Date`, then `self.modified` is Some?

File: src/suffix.rs [hunk 4]
```diff
@@ -209,15 +211,23 @@ impl SuffixScheme for AppendTimestamp {
...
         newest_suffix: Option<&TimestampSuffix>,
         suffix: &Option<TimestampSuffix>,
     ) -> io::Result<TimestampSuffix> {
         assert!(suffix.is_none());
         if suffix.is_none() {
-            let now = Local::now().format(self.format).to_string();
+            let mut now = Local::now();
+
+            match content_limit {
+                ContentLimit::Date => now = now - Duration::days(1),
```
Comment (Ploppz) [inline src/suffix.rs:223]: Why is this necessary? Just seems a bit ad hoc, I would try to think of a more general fix.

**Comment by jb-alvarado** [comment]
At the moment I have no better idea, then to subtract a day. I though we could use also the last modify date, but it work not always. For example when for multiple days was nothing written to the log, it takes an older date, that would be different to the linux logrotate behaviour.

File: examples/rotate_by_date.rs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+use file_rotate::{
+    compression::Compression, suffix::AppendTimestamp, suffix::FileLimit, ContentLimit, FileRotate,
+};
+use std::{io::Write};
```
Comment (BourgondAries) [inline examples/rotate_by_date.rs:4]: Unnecessary {}.

Please run cargo fmt and clippy on your commit.

File: src/lib.rs [hunk 6]
```diff
@@ -593,6 +616,18 @@ impl<S: SuffixScheme> Write for FileRotate<S> {
                     file.write_all(buf)?;
                 }
             }
+            ContentLimit::Date => {
+                let local = Local::now();
+                if let Some(modified) = self.modified {
+                    if local.date() > modified.date() {
```
Comment (BourgondAries) [inline src/lib.rs:622]: This seems like it would rotate when the day finishes.

**Review by BourgondAries — CHANGES_REQUESTED** [review]
This code needs tests and generalization. Afaict we now only rotate when the date (which has a resolution of a day) changes. It would instead be more useful to rotate upon changes given a "datetime mask" that informs us which differences are significant.

Also squash all the commits if they don't strictly need to be separate.

**Comment by Ploppz** [comment]
Maybe we don't need full generalization, considering that logrotate (because it is more or less the standard right now) only supports hourly, daily, weekly, monthly, yearly. Maybe a little more customization is nice (like how many times per day.. idk). But for example a "datetime mask" would not support "weekly".

**Comment by jb-alvarado** [comment]
I tried now, to make it more generic, with rotation options by:  hourly, daily, weekly, monthly, yearly, and add option for timestamp from yesterday, or hour before.

@BourgondAries, your idea of a "datetime mask" I don't now how to realise.

Sorry that my code is not so clean, I just started 3 month ago with rust. That is also the reason why I can not provide tests, because I don't know how to do this. @Ploppz you think you can take this part? :-).

File: src/lib.rs [hunk 7]
```diff
@@ -299,11 +300,28 @@ mod tests;
 
 // ---
 
+/// In which time frequency rotate the file.
```
Comment (BourgondAries) [inline src/lib.rs:303]: "At which frequency to rotate the file." perhaps

File: .gitignore [hunk 8]
```diff
@@ -1,2 +1,3 @@
 Cargo.lock
 target/
+logs/
```
Comment (BourgondAries) [inline .gitignore:3]: I think that lists like these in general should be kept sorted, so logs/ before target/

File: src/lib.rs [hunk 9]
```diff
@@ -299,11 +300,28 @@ mod tests;
...
+    /// Rotate yearly.
+    Yearly,
+}
+
 /// When to move files: Condition on which a file is rotated.
 #[derive(Clone, Debug)]
 pub enum ContentLimit {
     /// Cut the log at the exact size in bytes.
     Bytes(usize),
+    /// Cut the log at time base.
```
Comment (BourgondAries) [inline src/lib.rs:323]: Did you mean `timebase` instead of `time base`?

File: src/lib.rs [hunk 10]
```diff
@@ -381,6 +400,7 @@ impl<S: SuffixScheme> FileRotate<S> {
             ContentLimit::Bytes(bytes) => {
                 assert!(bytes > 0);
             }
+            ContentLimit::Time(_) => {}
```
Comment (BourgondAries) [inline src/lib.rs:403]: Seeing this, perhaps we should put the assertions inside the creation of the ContentLimit object (in a future commit) @Ploppz ?

Also I think the location of this item should be last in the struct since `Time` supercedes `Lines` lexicographically.

File: src/lib.rs [hunk 11]
```diff
@@ -431,6 +452,17 @@ impl<S: SuffixScheme> FileRotate<S> {
                                 self.count = 0;
                             }
                         }
+                        ContentLimit::Time(_) => {
+                            if let Ok(metadata) = file.metadata() {
+                                self.modified = match metadata.modified() {
```
Comment (BourgondAries) [inline src/lib.rs:457]: What happens on systems where `nomtime` (no modification time) is specified?

File: src/lib.rs [hunk 12]
```diff
@@ -593,6 +630,48 @@ impl<S: SuffixScheme> Write for FileRotate<S> {
...
+                                self.rotate()?;
+                            }
+                        }
+                        TimeFrequency::Monthly => {
+                            if local.format("%Y%m").to_string().parse::<i8>().unwrap_or(0)
+                                > modified
+                                    .format("%Y%m")
+                                    .to_string()
+                                    .parse::<i8>()
+                                    .unwrap_or(0)
```
Comment (BourgondAries) [inline src/lib.rs:658]: Why do we use %Y%m here? Afaict this gives `202204`, which is not a number that fits in `i8`.

File: src/lib.rs [hunk 13]
```diff
@@ -593,6 +630,48 @@ impl<S: SuffixScheme> Write for FileRotate<S> {
...
+                            {
+                                self.rotate()?;
+                            }
+                        }
+                        TimeFrequency::Yearly => {
+                            if local.year() > modified.year() {
```
Comment (BourgondAries) [inline src/lib.rs:664]: I like the clarity of this code!

File: src/suffix.rs [hunk 14]
```diff
@@ -214,10 +231,22 @@ impl SuffixScheme for AppendTimestamp {
...
         if suffix.is_none() {
-            let now = Local::now().format(self.format).to_string();
+            let mut now = Local::now();
+
+            match self.date_from {
+                DateFrom::DateYesterday => {
+                    now = now - Duration::days(1);
+                }
+                DateFrom::DateHourAgo => {
+                    now = now - Duration::hours(1);
```
Comment (BourgondAries) [inline src/suffix.rs:241]: Will this cause a time drift on the boundaries of a day (Say, during leap seconds)?

**Review by BourgondAries — COMMENTED** [review]
Mainly a few nitpicks but it's mostly looking great now. However, I do think it's essential to get some tests going here. @jb-alvarado we have some existing tests in this repository which you can use, in addition, I think you'll need to mock the `Local::now` function somehow to get deterministic tests.

File: src/lib.rs (hunk 10 above)
Comment (Ploppz) [reply src/lib.rs:403]: I like how currently you can construct ContentLimit directly without a function so I would leave it as it is

File: src/lib.rs (hunk 11 above)
Comment (jb-alvarado) [reply src/lib.rs:457]: Python is using mtime to, for log rotation, and it runs also on multiple platform. So I guess, the most common system will work. But maybe we found a fall-back?

File: src/lib.rs (hunk 12 above)
Comment (jb-alvarado) [reply src/lib.rs:658]: Ah, you are right, i8 is to small. I had no better idea to implement the right month. Because using only months will not work. When you change from December to January, you need the year info to.

File: src/suffix.rs (hunk 14 above)
Comment (jb-alvarado) [reply src/suffix.rs:241]: Hm, maybe. it looks like, that in python the always count with seconds: https://github.com/python/cpython/blob/7fce1063b6e5a366f8504e039a8ccdd6944625cd/Lib/logging/handlers.py#L232

**Comment by jb-alvarado** [comment]
> Mainly a few nitpicks but it's mostly looking great now. However, I do think it's essential to get some tests going here. @jb-alvarado we have some existing tests in this repository which you can use, in addition, I think you'll need to mock the `Local::now` function somehow to get deterministic tests.

I tried with libfaketime, but this don't work. Then I used touch, to modify the time from existing files. But yes something like mock I need. I only know this from python.

Later I will apply your suggestions.

**Comment by jb-alvarado** [comment]
I am a little undecided whether DateFrom is really correct. If the suffix has a timestamp with time, the result is a bit strange.

**Comment by jb-alvarado** [comment]
I tried to add tests. In this process I figure out also a issue in my code.

`self.modified` got only updated on init, or on rotation, but it have to be updated on every write operation.

**Comment by jb-alvarado** [comment]
I need to think a bit more about hourly and weekly rotation check. I don't think it works like that.

**Comment by jb-alvarado** [comment]
Was good to point out to the test! Now I think it should work. The DateTime mock is not so nice, but I don't know how to make it better.

**Comment by Ploppz** [comment]
Great job! The code to determine when to rotate now indeed looks much more correct. From a glance, test looks good.
One thing, run `cargo clippy --tests`, there are a few improvements possible there ;)

**Comment by jb-alvarado** [comment]
> One thing, run `cargo clippy --tests`, there are a few improvements possible there ;) If you wonder how I spotted them, it's my IDE (nvim with some plugins) that ran that command in the background automatically.

Thank you for the information. I was wondering already how you know this :-). I use vs code, it has also formatting and debugging options, but not with `cargo clippy` and `cargo fmt`. I need to look, if I find add-ons for that.
